### ISSUE TITLE
Inconsistent version id-s because of "current" flag

### ISSUE THREAD

**Issue by SimeonRolev** [OneDrive/onedrive-api-docs#1560]
Hi!
I'm working on an integration of OneDrive. When I run `https://graph.microsoft.com/v1.0/me/drive/items/<file_id>/versions`, the latest version is marked as "current". In the documentation example, all the versions have specific id-s (which is something I need).

I need to be able to consistently identify versions without them changing identifiers automatically. The issue with the "current" flag is that when you upload a new version, the new one becomes "current" and the older one gets an id like "107-70". I work on a product that uses file versions for multiple cloud providers and I need to save a real id of the file version, because I might need to reference it later. For example:

- User uploads `image.jpg` to OneDrive. Listing its `/versions` returns us an id of "current".
- User performs something with this version of the image on our site. So, we have to save which version of the file was used for that job in our DB - we save "current".
- The user uploads a new version of `image.jpg`. Listing `/versions` now applied the "current" id to a whole different version of the file. And gave a new version id for the used version - something like '107-70'.
- Now, when I need a `downloadUrl` for the first version, I can't get it, because I have no way to identify it no more - I have no proper id saved. When I used the file, it was "current" and I didn't know of '107-70', to be able to now get it out of the versions list.

I tried to use the `lastModifiedDateTime`, because it seems to be the only option for identifying, but it's very inconsistent and unreliable. More times than not, the `lastModifiedDateTime` for the file doesn't match any of the listed versions' `lastModifiedDateTime` field. It sometimes matches `fileSystemInfo.lastModifiedDateTime` or all the timestamps are different by milliseconds, seconds or even minutes. I don't really trust that path.

Another possible solution that I can think of is using the indexes of the `/versions` list at the time of execution - versions[0], versions[1] ... versions[<list_length_at_time_of_execution>]. But what happens if a version gets deleted, is it possible at all? I don't see such action on the website, restoring creates a new version. Do versions expire or get deleted automatically by OneDrive? What's the max limit of versions?

Bottom line, is there a way to consistently identify versions?
Thanks in advance!

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: fc7c4436-e84b-c5d0-d6c0-f89fa4087b29
* Version Independent ID: 8cc996e9-bc65-2e6a-cb2b-a74490572aad
* Content: [List versions of a file - OneDrive API - OneDrive dev center](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_list_versions?view=odsp-graph-online)
* Content Source: [docs/rest-api/api/driveitem_list_versions.md](https://github.com/OneDrive/onedrive-api-docs/blob/live/docs/rest-api/api/driveitem_list_versions.md)
* Product: **onedrive**
* GitHub Login: @JeremyKelley
* Microsoft Alias: **JeremyKe**

**Comment by ificator** [comment]
This has been an annoyance of mine as well - it's an identifier whose meaning changes on every upload, and it's incongruent with the SharePoint implementation. Unfortunately it's one of those cases where we cannot "fix" it because client may have a dependency on it. We could probably make a change that requires your client to opt-in to it though (e.g. via a `prefer` header)... would that work for you?

**Comment by ificator** [comment]
I took a look at this today and my knowledge of how versions are persisted was out-of-date. The short answer is it's no longer possible to know the id of the version until it's actually created. We could return a value other than `current` that would change on each upload, but you would not be able to use that value to query for the version when it's no longer the latest (if that makes sense). Here's a visualization:

*Original State*
|Version|
|-|
|107~70|
|106~70|
|104~70|

*State after uploading new content*
|Version|
|-|
|109~70|
|108~70|
|106~70|
|104~70|

Version `107~70` would not return anything after uploading new content.

Would this still help you? My concern is that it's even more confusing than `current`

**Comment by SimeonRolev** [comment]
Not sure if I'm missing something, but I can't see how that's helping with my issue, it's just a string change. As far as I understand, that still has the same issue like before - versions are not pointing to a resource consistently. Do I have an alternative for this?

**Comment by patrick-rodgers** [comment]
As part of a repository clean up effort we are closing older issues. If this issue remains, please: open a new issue, reference this issue, and provide any additional details that may help in resolution. Thank you for your understanding as we work to improve our responsiveness.
